contracts/external/cover/CoverOrder.sol

```diff
@@ -0,0 +1,307 @@
+    }
+
+    if (block.timestamp < executionDetails.notBefore) {
+      revert OrderCannotBeExecutedYet();
+    }
+
+    if (params.maxPremiumInAsset > executionDetails.maxPremiumInAsset) {
+      revert OrderPriceNotMet();
+    }
+
+    if(!memberRoles.checkRole(msg.sender, uint(IMemberRoles.Role.Member))) {
+      revert NotAMember();
+    }
+
+    if (params.owner == address(0) || params.owner == address(this)) {
+      revert InvalidOwnerAddress();
+    }
+
+    _verifySignature(params, executionDetails, signature);
+
+    // Ensure the order has not already been executed
+    if (orderStatus[signature] == OrderStatus.Executed) {
+      revert OrderAlreadyExecuted();
+    }
+
+    // Ensure the order is not cancelled
+    if (orderStatus[signature] == OrderStatus.Cancelled) {
+      revert OrderAlreadyCancelled();
+    }
+
+    // Mark the order as executed
+    orderStatus[signature] = OrderStatus.Executed;
+
+    // ETH payment
+    if (params.paymentAsset == ETH_ASSET_ID) {
+      coverId = _buyCoverEthPayment(params, poolAllocationRequests);
+    } else {
+      // ERC20 payment
+      coverId = _buyCoverErc20Payment(params, poolAllocationRequests);
+    }
+
+    // Emit event
+    emit OrderExecuted(params.owner, coverId, signature);
+
+  }
+
+  function cancelOrder(
+    BuyCoverParams calldata params,
+    ExecutionDetails calldata expirationDetails,
+    bytes calldata signature
+  ) external {
+
+    if (orderStatus[signature] == OrderStatus.Executed) {
+      revert OrderAlreadyExecuted();
+    }
+
+    if (params.owner != msg.sender) {
+      revert NotOrderOwner();
+    }
+
+    _verifySignature(params, expirationDetails, signature);
+
+    orderStatus[signature] = OrderStatus.Cancelled;
+    emit OrderCancelled(signature);
+  }
+
+  /// @notice Handles verification of the order signature
+  /// @param params Cover buy parameters
+  /// @param executionDetails Start and end date when the order can be executed and max premium in asset
+  /// @param signature The signature of the order
+  function _verifySignature(
+    BuyCoverParams calldata params,
+    ExecutionDetails calldata executionDetails,
+    bytes calldata signature
+  ) internal view {
+
+    // Hash the ExecutionDetails struct
+    bytes32 executionDetailsHash = keccak256(
+      abi.encode(
+        keccak256("ExecutionDetails(uint256 notBefore,uint256 deadline,uint256 maxPremiumInAsset)"),
+        executionDetails.notBefore,
+        executionDetails.deadline,
+        executionDetails.maxPremiumInAsset
+      )
+    );
+
+    // Hash the structured data
+    bytes32 structHash = keccak256(
+      abi.encode(
+        EXECUTE_ORDER_TYPEHASH,
+        params.productId,
+        params.amount,
+        params.period,
+        params.paymentAsset,
+        params.coverAsset,
+        params.owner,
+        executionDetailsHash
+      )
+    );
+
+    // Generate the digest (domain separator + struct hash)
+    bytes32 digest = _hashTypedDataV4(structHash);
```

Comment:
really tempted to use this digest as the order id and use it as the key in the `orderStatus` mapping. prevents potential signature malleability issues.